<!DOCTYPE HTML>
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>Genome Browser OpenCGA demo</title>

    <link rel="stylesheet" href="../css/genome-browser.css">

    <link rel="stylesheet" href="../../../node_modules/bootstrap/dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css">
    <link rel="stylesheet" href="../../../node_modules/qtip2/dist/jquery.qtip.min.css">
    <link rel="stylesheet" href="../../../styles/css/style.css">
    <link rel="stylesheet" href="../../../styles/css/toggle-switch.css">
    <link rel="stylesheet" href="../../../styles/css/magic-check.css">


    <script type="text/javascript" src="../../../node_modules/jquery/dist/jquery.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/bootstrap/dist/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/underscore/underscore-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/backbone/backbone-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/qtip2/dist/jquery.qtip.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/urijs/src/URI.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/cookies-js/dist/cookies.js"></script>
    <script type="text/javascript" src="../../../node_modules/crypto-js/crypto-js.js"></script>
</head>
<body>
    <div style="max-width:1400px;margin-left:auto;margin-right:auto;">
        <h1 style="font-weight:bold;">Genome Browser + OpenCGA Demo</h1>
        <hr>
        <div id="application"></div>
    </div>

    <script type="module">
        import {CellBaseClient} from "../../core/clients/cellbase/cellbase-client.js";
        import {OpenCGAClient} from "../../core/clients/opencga/opencga-client.js";
        import Region from "../../core/bioinfo/region.js";
        import GenomeBrowser from "../genome-browser.js";
        import GenomeBrowserConstants from "../genome-browser-constants.js";
        import GenomeBrowserUtils from "../genome-browser-utils.js";
        import GeneTrack from "../tracks/gene-track.js";
        import SequenceTrack from "../tracks/sequence-track.js";
        import GeneOverviewTrack from "../tracks/gene-overview-track.js";
        import OpenCGAVariantTrack from "../tracks/opencga-variant-track.js"

        import featuresOfInterest from "./examples/features-of-interest.js";
        
        // This is the parent element where GenomeBrowser will be rendered
        const parent = document.getElementById("application");

        // We first parse URL to check (and overwrite) if a CellBase Host and Version have been provided
        // if we do not provide the CellbaseHost then you will get those configured in the file "config.js"
        const params = new URLSearchParams(window.location.search);
        const cellBaseConfig = {
            host: params.get("CELLBASE_HOST") || "https://ws.zettagenomics.com/cellbase",
            version: params.get("CELLBASE_VERSION") || "v5",
            cache: {
                active: false,
            },
        };
        const cellBaseClient = new CellBaseClient(cellBaseConfig);

        // OpenCGA client config
        const opencgaClientConfig = {
            host: params.get("OPENCGA_HOST") || "https://ws.opencb.org/opencga-test",
            version: params.get("OPENCGA_VERSION") || "v2",
            // serverVersion: "1.4"
        };
        const opencgaClient = new OpenCGAClient(opencgaClientConfig);

        // Login into OpenCGA
        const openCGALogin = () => {
            const user = "demouser";
            const password = "demouser";
            return opencgaClient.login(user, password)
                .then(() => {
                    console.log(`User '${user}' successfully logged in in opencga.`);
                })
                .catch(error => {
                    console.log(`Error logging in '${user}'`, e);
                });
        };

        const run = availableSpecies => {
            const region = new Region({
                chromosome: "17",
                start: 43096757,
                end: 43112003,
            });
            const species = availableSpecies.vertebrates[0];

            const genomeBrowser = new GenomeBrowser(parent, {
                cellBaseClient: cellBaseClient,
                cellBaseHost: cellBaseConfig.host,
                cellBaseVersion: cellBaseConfig.version,
                width: parent.getBoundingClientRect().width,
                region: region,
                species: species,
                resizable: true,
                featuresOfInterest: featuresOfInterest,
            });

            // When GB is ready add tracks and draw
            genomeBrowser.on("ready", () => {
                // Overview tracks
                genomeBrowser.addOverviewTracks([
                    new GeneOverviewTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                ]);

                // Detail tracks
                genomeBrowser.addTracks([
                    // Sequence Track
                    new SequenceTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                    // Gene track
                    new GeneTrack({
                        title: "Gene",
                        cellBaseClient: cellBaseClient,
                    }),
                    // OpenCGA Variant track with query
                    new OpenCGAVariantTrack({
                        title: "Variant (Samples)",
                        closable: false,
                        minHistogramRegionSize: 20000000,
                        maxLabelRegionSize: 10000000,
                        minTranscriptRegionSize: 200000,
                        visibleRegionSize: 100000000,
                        height: 200,
                        opencgaClient: opencgaClient,
                        opencgaStudy: "demo@family:platinum",
                        query: {
                            sample: "NA12877,NA12878,NA12889,NA12890,NA12891,NA12892",
                        },
                        highlights: [
                            {
                                id: "test1",
                                name: "[TEST] Length > 0",
                                description: "This variant has length > 0.",
                                condition: feature => feature.end - feature.start + 1 > 0,
                            },
                        ],
                    }),
                ]);

                genomeBrowser.draw();
            });
        }

        const promisesList = [
            openCGALogin(),
            GenomeBrowserUtils.getSpeciesFromCellBase(cellBaseClient),
        ];

        Promise.all(promisesList).then(values => {
            // values[0] -> response from openCGALogin
            // values[1] -> response from getSpeciesFromCellBase
            return run(values[1]);
        });

    </script>
</body>
</html>
