<!DOCTYPE HTML>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>Genome Browser demo</title>

    <link rel="stylesheet" href="../css/genome-browser.css">

    <link rel="stylesheet" href="../../../node_modules/bootstrap/dist/css/bootstrap.min.css">
    <link rel="stylesheet" href="../../../node_modules/@fortawesome/fontawesome-free/css/all.min.css">
    <link rel="stylesheet" href="../../../node_modules/qtip2/dist/jquery.qtip.min.css">
    <link rel="stylesheet" href="../../../styles/css/style.css">
    <link rel="stylesheet" href="../../../styles/css/toggle-switch.css">
    <link rel="stylesheet" href="../../../styles/css/magic-check.css">

    <script type="text/javascript" src="../../../node_modules/jquery/dist/jquery.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/bootstrap/dist/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/underscore/underscore-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/backbone/backbone-min.js"></script>
    <script type="text/javascript" src="../../../node_modules/qtip2/dist/jquery.qtip.min.js"></script>
    <script type="text/javascript" src="../../../node_modules/cookies-js/dist/cookies.js"></script>
    <script type="text/javascript" src="../../../node_modules/crypto-js/crypto-js.js"></script>
</head>
<body>
    <div style="max-width:1400px;margin-left:auto;margin-right:auto;">
        <h1 style="font-weight:bold;">Genome Browser Demo</h1>
        <hr>
        <div id="application"></div>
    </div>

    <script type="module">
        import {CellBaseClient} from "../../core/clients/cellbase/cellbase-client.js";
        import Region from "../../core/bioinfo/region.js";
        import Utils from "../../core/utils.js";
        import GenomeBrowser from "../genome-browser.js";
        import GenomeBrowserConstants from "../genome-browser-constants.js";
        import GenomeBrowserUtils from "../genome-browser-utils.js";
        import GeneTrack from "../tracks/gene-track.js";
        import VariantTrack from "../tracks/variant-track.js";
        import SequenceTrack from "../tracks/sequence-track.js";
        import GeneOverviewTrack from "../tracks/gene-overview-track.js";

        import featuresOfInterest from "./examples/features-of-interest.js";

        // We first parse URL to check (and overwrite) if a CellBase Host and Version have been provided
        // if we do not provide the CellbaseHost then you will get those configured in the file "config.js"
        const params = new URLSearchParams(window.location.search);
        const cellBaseConfig = {
            host: params.get("CELLBASE_HOST") || "https://ws.zettagenomics.com/cellbase",
            version: params.get("CELLBASE_VERSION") || "v5",
            cache: {
                active: false,
            },
        };
        const cellBaseClient = new CellBaseClient(cellBaseConfig);

        const parent = document.getElementById("application");
        const region = new Region({
            chromosome: "13",
            start: 32996311,
            end: 32996450,
        });

        const run = availableSpecies => {
            const species = availableSpecies.vertebrates[0];
            const genomeBrowser = new GenomeBrowser(parent, {
                cellBaseClient: cellBaseClient,
                cellBaseHost: cellBaseConfig.host,
                cellBaseVersion: cellBaseConfig.version,
                width: parent.getBoundingClientRect().width,
                region: region,
                species: species,
                resizable: true,
                featuresOfInterest: featuresOfInterest,
            });

            // When GB is ready add tracks and draw
            genomeBrowser.on("ready", () => {
                // Overview tracks
                genomeBrowser.addOverviewTracks([
                    new GeneOverviewTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                ]);

                // Detail tracks
                genomeBrowser.addTracks([
                    // Sequence Track
                    new SequenceTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                    // Gene track
                    new GeneTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                    // CellBase Variant track
                    new VariantTrack({
                        cellBaseClient: cellBaseClient,
                    }),
                ]);

                genomeBrowser.draw();
            });
        }

        GenomeBrowserUtils.getSpeciesFromCellBase(cellBaseClient)
            .then(availableSpecies => {
                run(availableSpecies)
            });

    </script>
</body>
</html>
